Vim basics - blog post (#194)
ombulabs/blog: _posts/2019-07-29-vim-basics.markdown

---
layout: post
title: vim Basics
date: 2019-07-29 22:28:02
categories:
author: bronzdoc
---

# VIM basics

I would not bore you by telling you the history of vim, and the differences between vi and vim and all of its unix philosophy....

> If you are interested in that, you can give it a look [here](https://en.wikipedia.org/wiki/Vim_(text_editor).

Reviewer: this link is not working (not correctly formatted), you probably need to remove a parentheses

> So all you need to know is that vim is a text editor... in my opinion one of vim strengths comes with how flexible it can be. It can be as simple or complex as you wish. It can serve you to edit a simple config file on a server or you can make it a complete code editing environment if you want.

Reviewer: ```suggestion
So all you need to know is that vim is a text editor. In my opinion one of vim's strengths comes with how flexible it can be. It can be as simple or complex as you wish. It can be used to edit a simple config file on a server or you can make it a complete code editing environment.
```

> In this blog post I'll show you the basics of vim and the smallest amount of knowledge to make vim useful for yourself and yes I'll teach you how to exit ;)

Reviewer: ```suggestion
In this blog post I'll show you the basics of vim and provide you with the smallest amount of knowledge necessary to make vim useful for yourself, and yes I'll teach you how to exit ;)
```

> I'll asume you already have vim installed, if you are on a \*nix system you proably already have it.

Reviewer: ```suggestion
I'll assume you already have vim installed, if you are on a \*nix system you probably already have it.
```

Open terminal and type: `vim vim-basic`.

> You'll see something like the folowing:

Reviewer: ```suggestion
You'll see something like the following:
```

> ```

Reviewer: The formatting is off here. The code block doesn't end in the right place. See screenshot. 

<img width="880" alt="Screen Shot 2019-08-05 at 3 29 16 PM" src="https://user-images.githubusercontent.com/28606034/62489856-d864b700-b795-11e9-91c7-b370f22ae65c.png">

                                                      VIM - Vi IMproved

                                                       version 8.1.500
                                                   by Bram Moolenaar et al.
                                         Vim is open source and freely distributable

                                                Help poor children in Uganda!
                                        type  :help iccf<Enter>       for information

                                        type  :q<Enter>               to exit
                                        type  :help<Enter>  or  <F1>  for on-line help
                                        type  :help version8<Enter>   for version info

                                                Running in Vi compatible mode
                                        type  :set nocp<Enter>        for Vim defaults
                                        type  :help cp-default<Enter> for info on this
```

congrats! you are using vim ha!

## VIM modes

> Vim is a "modal editor" that means vim will behave different depending on which mode you currently are.

Reviewer: ```suggestion
Vim is a "modal editor", which means vim will behave different depending on which mode you are currently on.
```

> I'll introduce you to three basic modes. **NORMAL**, **INSERT** and **VISUAL**. There are a couple more, but this are the basics ones, and it's all you need to know for now.

Reviewer: ```suggestion
I'll introduce you to three basic modes. **NORMAL**, **INSERT** and **VISUAL**. There are a couple more, but these are the basics ones, and they are all you need to know for now.
```

> ## NORMAL - The default's vim mode

Reviewer: ```suggestion
## NORMAL - The default vim mode
```

This is where you'll spend most of your time and it's the mode where the magic happens. In this mode you can execute commands.
Commands can be a simple letter, for example if you press the letter `j` you'll see your cursor goes down one line and if you press the letter `k` it will go up.

> In this mode you'll do things like navigate trough your text file, removing, copyieng and pasting text.

Reviewer: ```suggestion
In this mode you'll do things like navigate through your text file, removing, copying and pasting text.
```

### Navigation

> While you are in **NORMAL** mode you would need to move around your text file. This are the basics commads that will help you achieve that.

Reviewer: ```suggestion
While you are in **NORMAL** mode you would need to move around your text file. These are the basics commands that will help you achieve that.
```

> ```

Reviewer: Formatting issues here.

Reviewer: <img width="938" alt="Screen Shot 2019-08-05 at 3 32 48 PM" src="https://user-images.githubusercontent.com/28606034/62490006-4d37f100-b796-11e9-8146-0bc5a27228ed.png">

      k
      ^
  h < + > l
      v
      j
```

### Removing text
  dd - delete the current line.

### Copy text
  yy - copy(yank) the current line.

### Pasting text

>   p  - Paste to the underneath line.

Reviewer: ```suggestion
  p  - paste to the line underneath.
```

## INSERT - The mode where you can "insert" text

>   You problay were wondering, how do I actually write something into vim, well... this is the mode where you can do that.

Reviewer: ```suggestion
  You probably were wondering, how do I actually write something into vim, well... this is the mode where you can do that.
```

>   You enter INSERT mode by pressing the letter `i`. In this mode you'll use you keyboard like in any other editor, vim will not behave special when you are in this mode.

Reviewer: ```suggestion
  You enter INSERT mode by pressing the letter `i`. In this mode you'll use your keyboard like in any other editor, vim will not behave in a special way when you are in this mode.
```

  In vim, you are not expected to be in this mode all the time Whenever you finish your writing you go back to **NORMAL** mode, you do that by pressing the **ESC** key.

  After writing or editing your changes you'll probably want to save your file and quit vim.
  Press **ESC** and go back to normal mode, then you could do the any of following:

  * :w  - Save your changes.
  * :q  - Quit vim.

>   * :q! - Quit vim even if your changes where not saved.

Reviewer: ```suggestion
  * :q! - Quit vim even if your changes were not saved.
```

## VISUAL - The mode in which you can select chunks of text
 While in normal mode you enter this mode by pressing the letter `v`. You'll notice that the text will highlight if you go up, down, left or right.

>  You could select a paragraph, a single word or even a letter and apply commands to that selected piece of text. You could copy(yy) a selected of text or delete (dd) it if you wish. to exit visual mode you press `ESC` and you will return to **NORMAL** mode.

Reviewer: ```suggestion
 You could select a paragraph, a single word or even a letter and apply commands to that selected piece of text. You could copy(yy) a selection of text or delete (dd) it if you wish. To exit visual mode you press `ESC` and will return to **NORMAL** mode.
```

## Resources

>  This are a couple of resources that can help you improve your vim skilss.

Reviewer: ```suggestion
 These are a couple of resources that can help you improve your vim skills.
```

  * vimtutor                   - If you are on a \*nix system you can type in the terminal `vimtutor` for a vim tutorial.
  * http://vim-adventures.com/ - A game that will teach you vim

>   * :h :help                   - vim goto help command

Reviewer: ```suggestion
  * :h :help                   - vim go to help command
```

## Conclusion

>  This is a really basic introduction to vim and it's just the top of the iceberg of what vim really can do. I hope this little introduction will motivate you to keep learning vim and showed you that to start using vim you only need a couple of concepts and commands.

Reviewer: ```suggestion
 This is a really basic introduction to vim and it's just the tip of the iceberg of what vim really can do. I hope this little introduction will motivate you to keep learning vim and showed you that to start using vim you only need a couple of concepts and commands.
```

---
layout: post

> title: vim Basics

Reviewer 2: @bronzdoc Maybe you can change the title to be Vim Guide for Beginners, or something like that?

Author: Sounds good 👍

> date: 2019-07-29 22:28:02

Author: Done @lubc

categories:
author: bronzdoc
---

# VIM basics

> I would not bore you by telling you the history of vim, and the differences between vi and vim and all of its philosophy...

Reviewer: ```suggestion
I will not bore you by telling you the history of vim, and the differences between vi and vim and all of its philosophy...
```

> I would not bore you by telling you the history of vim, and the differences between vi and vim and all of its philosophy...

Reviewer 2: @bronzdoc I think you should capitalize the `V` every time you say vim
```suggestion
I would not bore you by telling you the history of Vim, and the differences between vi and Vim and all of its philosophy...
```

Reviewer 2: @bronzdoc Also, add a link to the official page or github repo the first time you mention it

> s

Reviewer: Random "s" here?

If you are interested in that, you can give it a look [here](https://en.wikipedia.org/wiki/Vim_(text_editor)).

<!--more-->

So all you need to know is that vim is a text editor. In my opinion one of vim's strengths comes with how flexible it can be. It can be as simple or complex as you wish. It can be used to edit a simple config file on a server or you can make it a complete code editing environment.

In this blog post I'll show you the basics of vim and provide you with the smallest amount of knowledge necessary to make vim useful for yourself, and yes I'll teach you how to exit ;)

> I'll assume you already have vim installed, if you are on a \*nix system you probably already have it.

Reviewer 2: @bronzdoc Since this is a beginner guide, maybe you can let them know how to install it and how to check if they have it already.

> I'll assume you already have vim installed, if you are on a \*nix system you probably already have it.

Author: ```
*nix means UNIX-like; it is an operating system that behaves in a manner similar to that of a UNIX operating system without necessarily conforming to the Single UNIX Specification.
```
I think *nix is appropiate here

> I'll assume you already have vim installed, if you are on a \*nix system you probably already have it.

Author: I think i have never explicitly installed vim in my life ha! but for correctness i think is a good idea.

> I'll assume you already have vim installed, if you are on a \*nix system you probably already have it.

Author: Done @lubc

Open terminal and type: `vim vim-basic`.

You'll see something like the following:
```
                                              VIM - Vi IMproved

                                               version 8.1.500
                                           by Bram Moolenaar et al.
                                 Vim is open source and freely distributable

                                        Help poor children in Uganda!
                                type  :help iccf<Enter>       for information

                                type  :q<Enter>               to exit
                                type  :help<Enter>  or  <F1>  for on-line help
                                type  :help version8<Enter>   for version info

                                        Running in Vi compatible mode
                                type  :set nocp<Enter>        for Vim defaults
                                type  :help cp-default<Enter> for info on this

> ```

Reviewer 2: @bronzdoc I don't see this when I type that command. I see a blank new file

Author: You are correct @lubc 

I'm so used to type `vim` and then operate from there that i forgot that `vim filename` would not show that text. 🤦‍♂

congrats! you are using vim ha!

## VIM modes

> Vim is a "modal editor", which means vim will behave different depending on which mode you are currently on.

Reviewer: ```suggestion
Vim is a "modal editor", which means vim will behave differently depending on which mode you are currently on.
```

I'll introduce you to three basic modes. **NORMAL**, **INSERT** and **VISUAL**. There are a couple more, but these are the basics ones, and they are all you need to know for now.

## NORMAL - The default vim mode

This is where you'll spend most of your time and it's the mode where the magic happens. In this mode you can execute commands.
Commands can be a simple letter, for example if you press the letter `j` you'll see your cursor goes down one line and if you press the letter `k` it will go up.
In this mode you'll do things like navigate through your text file, removing, copying and pasting text.

### Navigation

> While you are in **NORMAL** mode you would need to move around your text file. These are the basics commands that will help you achieve that.

Reviewer 2: @bronzdoc I guess you can mention that you can achieve the same with the arrow keys, right?

Author: @lubc If you are using vim you should not use the arrow keys,  I'll not teach such devil practice 🤓

```
      k
      ^
  h < + > l

>       v

Reviewer 2: @bronzdoc is the `v` correct here?

Author: Yes, I could not find any other character to represent a down arrow and even vimtutor uses it to represent a down arrow.

      j
```

### Removing text
  dd - delete the current line.

### Copy text
  yy - copy(yank) the current line.

### Pasting text
  p  - paste to the line underneath.

## INSERT - The mode where you can "insert" text
  You probably were wondering, how do I actually write something into vim, well... this is the mode where you can do that.

  You enter INSERT mode by pressing the letter `i`. In this mode you'll use your keyboard like in any other editor, vim will not behave in a special way when you are in this mode.

  In vim, you are not expected to be in this mode all the time Whenever you finish your writing you go back to **NORMAL** mode, you do that by pressing the **ESC** key.

  After writing or editing your changes you'll probably want to save your file and quit vim.
  Press **ESC** and go back to normal mode, then you could do the any of following:

  * :w  - Save your changes.
  * :q  - Quit vim.

>   * :q! - Quit vim even if your changes were not saved.

Reviewer 2: @bronzdoc It will be useful to mention `:wq` too

Author: 🤔 `:wq` is the combination of two of the commands i'm showing

## VISUAL - The mode in which you can select chunks of text

>  While in normal mode you enter this mode by pressing the letter `v`. You'll notice that the text will highlight if you go up, down, left or right.

Reviewer 2: @bronzdoc To be clear:
```suggestion
 While in normal mode you enter the visual mode by pressing the letter `v`. You'll notice that the text will highlight if you go up, down, left or right.
```
